rbac/management/role/relation_api_dual_write_handler.py

```diff
@@ -34,31 +41,154 @@ class DualWriteException(Exception):
+        add_json = []
+        for relation in relations_to_add:
+            add_json.append(relationship_to_json(relation))
+
+        remove_json = []
+        for relation in relations_to_remove:
+            remove_json.append(relationship_to_json(relation))
+
+        replication_event = {"relations_to_add": add_json, "relations_to_remove": remove_json}
+        return replication_event
+
+    def _save_replication_event(self, payload, event_type, aggregateid):
+        """Save replication event."""
+        logger.info(
+            "[Dual Write] Save replication event into outbox table for role(%s): '%s'", self.role.uuid, self.role.name
+        )
+        logger.info("[Dual Write] Replication event: %s for role(%s): '%s'", payload, self.role.uuid, self.role.name)
+        # https://debezium.io/documentation/reference/stable/transformations/outbox-event-router.html#basic-outbox-table
+        outbox_record = Outbox.objects.create(
+            aggregatetype="RelationReplicationEvent",
+            aggregateid=aggregateid,
+            event_type=event_type,
+            payload=payload,
+        )
+        outbox_record.delete()
+
+
+class NoopReplicator(RelationReplicator):
+    """Noop replicator."""
+
+    def replicate(self, event: ReplicationEvent):
+        """Noop."""
+        pass
+
+
 class RelationApiDualWriteHandler:
     """Class to handle Dual Write API related operations."""
 
-    def __init__(self, role, event_type):
+    @classmethod
+    def for_system_role_event(
+        cls,
+        role: Role,
+        tenant: Tenant,
+        event_type: ReplicationEventType,
+        replicator: Optional[RelationReplicator] = None,
+    ):
+        """Create a RelationApiDualWriteHandler for a system role."""
+        return cls(role, event_type, replicator, tenant)
+
+    def __init__(
+        self,
+        role: Role,
+        event_type: ReplicationEventType,
+        replicator: Optional[RelationReplicator] = None,
+        tenant: Optional[Tenant] = None,
+    ):
         """Initialize RelationApiDualWriteHandler."""
         if not self.replication_enabled():
+            self._replicator = NoopReplicator()
             return
         try:
-            self.role_relations = []
-            self.current_role_relations = []
+            self._replicator = replicator if replicator else OutboxReplicator(role)
+            self.event_type = event_type
+            self.role_relations: list[common_pb2.Relationship] = []
+            self.current_role_relations: list[common_pb2.Relationship] = []
             self.role = role
             self.binding_mapping = None
-            self.tenant_id = role.tenant_id
-            self.org_id = role.tenant.org_id
-            self.event_type = event_type
+
+            binding_tenant = tenant if tenant is not None else role.tenant
+
+            if binding_tenant.name == "public":
```

Comment:
tenant has column `tenant_name` instead of `name`

Reply:
Ty!